### PR TITLE
fuse constant shape for reshape

### PR DESCRIPTION
**Pull request by HSQ79815**
Signed-off-by: haoshengqiang <haoshengqiang79@163.com>

### CODE REVIEW

File: onnxoptimizer/passes/fuse_concat_shape_for_reshape.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    std::vector<int64_t> shapes;
+    for (const auto *v : concat->inputs()) {
+      const uint32_t kind = v->node()->kind();
+      const Tensor *tensor = nullptr;
+      if (kind == kConstant) {
+        tensor = &v->node()->t(kvalue);
+      } else if (kind == kParam &&
+                 std::find(initializer_names.cbegin(), initializer_names.cend(),
+                           v->uniqueName()) != initializer_names.cend()) {
```
Comment (daquexian) [inline onnxoptimizer/passes/fuse_concat_shape_for_reshape.h:44]: is this condition necessary?

File: onnxoptimizer/passes/fuse_concat_shape_for_reshape.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        if (v->sizes().size() != 1) {
+          return false;
+        }
+        shapes.push_back(-1);
+        continue;
+      }
+      const auto data = ParseData<int64_t>(tensor);
+      std::copy(data.cbegin(), data.cend(), std::back_inserter(shapes));
+    }
+    int unkown_dim_count = 0;
```
Comment (daquexian) [inline onnxoptimizer/passes/fuse_concat_shape_for_reshape.h:56]:
```suggestion
    int unknown_dim_count = 0;
```

File: onnxoptimizer/passes/fuse_concat_shape_for_reshape.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                           v->uniqueName()) != initializer_names.cend()) {
+        tensor = &*graph.getInitializer(v->uniqueName());
+      } else {
+        if (v->sizes().size() != 1) {
+          return false;
+        }
+        shapes.push_back(-1);
+        continue;
+      }
+      const auto data = ParseData<int64_t>(tensor);
```
Comment (daquexian) [inline onnxoptimizer/passes/fuse_concat_shape_for_reshape.h:53]:
```suggestion
      if (tensor->elem_type() != TensorProto::INT64) {
        return false;
      }
      const auto data = ParseData<int64_t>(tensor);
```

File: onnxoptimizer/passes/fuse_concat_shape_for_reshape.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+namespace ONNX_NAMESPACE {
+namespace optimization {
+
+struct FuseConcatShapeForReshape final : public PredicateBasedPass {
+  explicit FuseConcatShapeForReshape()
+      : PredicateBasedPass(PassType::Fuse, PassEfficiency::Complete,
+                           PassOptimizationType::Compute) {}
+
+  std::string getPassName() const override {
+    return "fuse_concat_shape_for_reshape";
```
Comment (daquexian) [inline onnxoptimizer/passes/fuse_concat_shape_for_reshape.h:22]: how about "fuse_concat_and_reshape"

**Review by daquexian — APPROVED** [review]

File: onnxoptimizer/passes/fuse_concat_shape_for_reshape.h (hunk 1 above)
Comment (HSQ79815) [reply onnxoptimizer/passes/fuse_concat_shape_for_reshape.h:44]: tensor in kParam maybe a input

File: onnxoptimizer/passes/fuse_concat_shape_for_reshape.h (hunk 4 above)
Comment (HSQ79815) [reply onnxoptimizer/passes/fuse_concat_shape_for_reshape.h:22]: good

**Pull request merged**